File: rabbitmq_protobufs_android/app/src/main/java/com/hidepixel/rabbitmq_protobufs_android/MainActivity.java
package com.hidepixel.rabbitmq_protobufs_android;

import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.Button;

import com.rabbitmq.client.AMQP;
import com.rabbitmq.client.Channel;
import com.rabbitmq.client.Connection;
import com.rabbitmq.client.ConnectionFactory;
import com.rabbitmq.client.Consumer;
import com.rabbitmq.client.DefaultConsumer;
import com.rabbitmq.client.Envelope;

import java.io.IOException;
import java.util.concurrent.TimeoutException;

public class MainActivity extends AppCompatActivity {

    private static final String PROTO_QUEUE_NAME = "proto";
    private static final String JSON_QUEUE_NAME = "json";


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);


        Button button = (Button) findViewById(R.id.button);
        Button button2 = (Button) findViewById(R.id.button2);

        button.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                new Thread(new Runnable() {
                    @Override
                    public void run() {
                        try {
                            getProtobufMessageFromRabbit();
                            getJSONMessageFromRabbit();
                        } catch (IOException | TimeoutException e) {
                            e.printStackTrace();
                        }
                    }
                }).start();
            }
        });

        button2.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                new Thread(new Runnable() {
                    @Override
                    public void run() {
                        try {
                            sendProtobufMessageToRabbit();
                            sendJSONMessageToRabbit();
                        } catch (IOException | InterruptedException | TimeoutException e) {
                            e.printStackTrace();
                        }
                    }
                }).start();
            }
        });
    }

    public void sendJSONMessageToRabbit() throws IOException, TimeoutException, InterruptedException {

        ConnectionFactory factory = new ConnectionFactory();
        factory.setHost("10.0.2.2");
        Connection connection = factory.newConnection();
        Channel channel = connection.createChannel();

        channel.queueDeclare(JSON_QUEUE_NAME, false, false, false, null);


        String json = "id: '56fa595034b950a97a63e3a0', " +
                "updated_at: 1459247440, " +
                "created_at: 1459247440, " +
                "web_site: 'http://datsite.com/', " +
                "description: 'Gotta love some description', " +
                "name: 'Say my name', " +
                "version: 0, " +
                "picture: 'applications/56fa595034b950a97a63e3a0/086eda82-3d08-4b4d-ba38-2596d0bf6a98.jpeg'";

        channel.basicPublish("", JSON_QUEUE_NAME, null, json.getBytes());

        channel.close();
        connection.close();
    }

    public void sendProtobufMessageToRabbit() throws IOException, TimeoutException, InterruptedException {

        ConnectionFactory factory = new ConnectionFactory();
        factory.setHost("10.0.2.2"); // localhost equivalent for the android emulator
        Connection connection = factory.newConnection();
        Channel channel = connection.createChannel();

        channel.queueDeclare(PROTO_QUEUE_NAME, false, false, false, null);

        PartnerProto.Partner message = PartnerProto.Partner.newBuilder()
                .setId("56fa595034b950a97a63e3a0")
                .setUpdatedAt(1459247440)
                .setCreatedAt(1459247440)
                .setWebSite("http://datsite.com/")
                .setDescription("Gotta love some description")
                .setName("Say my name")
                .setVersion(0)
                .setPicture("applications/56fa595034b950a97a63e3a0/086eda82-3d08-4b4d-ba38-2596d0bf6a98.jpeg")
                .build();

        channel.basicPublish("", PROTO_QUEUE_NAME, null, message.toByteArray());

        channel.close();
        connection.close();
    }

    public void getProtobufMessageFromRabbit() throws IOException, TimeoutException {

        ConnectionFactory connectionFactory = new ConnectionFactory();
        connectionFactory.setHost("10.0.2.2");
        Connection connection = connectionFactory.newConnection();
        Channel channel = connection.createChannel();

        channel.queueDeclare(PROTO_QUEUE_NAME, false, false, false, null);

        Consumer consumer = new DefaultConsumer(channel) {
            @Override
            public void handleDelivery(String consumerTag, Envelope envelope, AMQP.BasicProperties properties, byte[] body) throws IOException {
                PartnerProto.Partner proto = PartnerProto.Partner.parseFrom(body);


                Log.v("rabbitmq receive", "Received protobuf '" + proto.toString() + "'");
            }
        };

        channel.basicConsume(PROTO_QUEUE_NAME, true, consumer);
    }

    public void getJSONMessageFromRabbit() throws IOException, TimeoutException {
        ConnectionFactory connectionFactory = new ConnectionFactory();
        connectionFactory.setHost("10.0.2.2");
        Connection connection = connectionFactory.newConnection();
        Channel channel = connection.createChannel();

        channel.queueDeclare(JSON_QUEUE_NAME, false, false, false, null);

        Consumer consumer = new DefaultConsumer(channel) {
            @Override
            public void handleDelivery(String consumerTag, Envelope envelope, AMQP.BasicProperties properties, byte[] body) throws IOException {
                String jsonMessage = new String(body, "UTF-8");

                Log.v("rabbitmq receive", "Received json '" + jsonMessage + "'");
            }
        };

        channel.basicConsume(JSON_QUEUE_NAME, true, consumer);
    }

}

File: RabbitMQAndProtobufs/app/src/main/java/com/rcosteira/rabbitmqandprotobufs/MainActivity.kt
package com.rcosteira.rabbitmqandprotobufs

import android.os.Bundle
import android.util.Log
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.Observer
import com.google.android.material.button.MaterialButton
import com.rabbitmq.client.*
import com.rcosteira.rabbitmq_protobufs_android.UserProto
import kotlinx.coroutines.*
import kotlin.coroutines.CoroutineContext

class MainActivity : AppCompatActivity(), CoroutineScope {

    companion object {
        const val PROTO_FROM_SERVER = "proto_from_server"
        const val JSON_FROM_SERVER = "json_from_server"
        const val PROTO_FROM_ANDROID = "proto_from_android"
        const val JSON_FROM_ANDROID = "json_from_android"
    }

    lateinit var receivedMessageTextView: TextView
    lateinit var connection: Connection
    lateinit var channel: Channel


    private lateinit var job: Job
    override val coroutineContext: CoroutineContext
        get() = job + Dispatchers.Main

    private val handler = CoroutineExceptionHandler { _, throwable ->
        Log.e("Exception", ":$throwable")
    }

    private val incomingMessage: MutableLiveData<String> = MutableLiveData()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        job = SupervisorJob()

        setupMessageSendingButtons()

        setupReceivedMessageTextView()

        launch(handler) {
            withContext(Dispatchers.Default) {
                setupRabbitConnection()
                listenToIncomingMessages()
            }

        }
    }

    private fun setupRabbitConnection() {
        connection = createConnection()
        channel = connection.createChannel()
        channel.queueDeclare(JSON_FROM_SERVER, false, false, false, null)
        channel.queueDeclare(PROTO_FROM_SERVER, false, false, false, null)
        channel.queueDeclare(JSON_FROM_ANDROID, false, false, false, null)
        channel.queueDeclare(PROTO_FROM_ANDROID, false, false, false, null)
    }

    private fun createConnection(): Connection {
        val connectionFactory = getConnectionFactory()
        return connectionFactory.newConnection()
    }

    private fun getConnectionFactory(): ConnectionFactory {
        val connectionFactory = ConnectionFactory()
        connectionFactory.host = "10.0.2.2" // localhost equivalent for the android emulator

        return connectionFactory
    }

    private fun setupMessageSendingButtons() {
        val protoButton: MaterialButton = findViewById(R.id.protoButton)
        val jsonButton: MaterialButton = findViewById(R.id.jsonButton)

        protoButton.setOnClickListener {
            sendMessage(PROTO_FROM_ANDROID, ::getProtobufMessage)
        }

        jsonButton.setOnClickListener {
            sendMessage(JSON_FROM_ANDROID, ::getJsonMessage)
        }
    }

    private fun sendMessage(queueName: String, formatHandler: () -> ByteArray) {
        val message = formatHandler()

        launch(handler) {
            withContext(Dispatchers.Default) {
                channel.basicPublish("", queueName, null, message)
            }
        }
    }


    private fun getProtobufMessage(): ByteArray {
        val phones = listOf(
            UserProto.User.PhoneNumber.newBuilder()
                .setNumber("+123456789")
                .setType(UserProto.User.PhoneType.HOME)
                .build(),
            UserProto.User.PhoneNumber.newBuilder()
                .setNumber("+987654321")
                .setType(UserProto.User.PhoneType.MOBILE)
                .build()
        )

        val user = UserProto.User.newBuilder()
            .setId("56fa595034b950a97a63e3a0")
            .setUpdatedAt(System.currentTimeMillis())
            .setCreatedAt(System.currentTimeMillis())
            .setName("Barry Allen")
            .setVersion(1)
            .setPicture("resources/New Suit.jpeg")
            .addAllPhones(phones)
            .build()

        return user.toByteArray()
    }

    private fun getJsonMessage(): ByteArray {
        val json = "{" +
                "\"User\": {" +
                "\"id\": 56fa595034b950a97a63e3a0," +
                "\"updated_at\": \"${System.currentTimeMillis()}\", " +
                "\"created_at\": \"${System.currentTimeMillis()}\", " +
                "\"name\": \"Barry Allen\"," +
                "\"version\": \"1," +
                "\"picture\": \"resources/New Suit.jpeg\"," +
                "\"phones\": [" +
                "{" +
                "\"number\": \"+123456789\"," +
                "\"type\": \"HOME\"" +
                "}," +
                "{" +
                "\"number\": \"+987654321\"," +
                "\"type\": \"MOBILE\"" +
                "}" +
                "]" +
                "}" +
                "}"

        return json.toByteArray()
    }

    private fun setupReceivedMessageTextView() {
        receivedMessageTextView = findViewById(R.id.receivedMessageTextView)
        incomingMessage.observe(this, Observer {
            receivedMessageTextView.text = it
        })
    }

    private fun listenToIncomingMessages() {
        listenToProtoQueue()
        listenToJsonQueue()
    }

    private fun listenToProtoQueue() {
        val protoConsumer = getProtoConsumer()
        channel.basicConsume(PROTO_FROM_SERVER, true, protoConsumer)
    }

    private fun getProtoConsumer(): Consumer {
        return object : DefaultConsumer(channel) {
            override fun handleDelivery(
                consumerTag: String?,
                envelope: Envelope?,
                properties: AMQP.BasicProperties?,
                body: ByteArray?
            ) {
                body?.let {
                    Log.d("MainActivity", "Got a protobuf")
                    val user = UserProto.User.parseFrom(it)
                    incomingMessage.postValue(user.toString())
                }
            }
        }
    }

    private fun listenToJsonQueue() {
        val jsonConsumer = getJsonConsumer()
        channel.basicConsume(JSON_FROM_SERVER, true, jsonConsumer)
    }

    private fun getJsonConsumer(): Consumer {
        return object : DefaultConsumer(channel) {
            override fun handleDelivery(
                consumerTag: String?,
                envelope: Envelope?,
                properties: AMQP.BasicProperties?,
                body: ByteArray?
            ) {
                body?.let {
                    Log.d("MainActivity", "Got a json")
                    incomingMessage.postValue(String(it, charset("UTF-8")))
                }
            }
        }
    }

    override fun onDestroy() {

        connection.close()
        channel.close()

        coroutineContext.cancelChildren()

        super.onDestroy()
    }
}